### DISCUSSION TITLE
Less hacky way to build `FROM SCRATCH` dockers?

### DISCUSSION THREAD

**Discussion by jchvz** [dagger/dagger#2517] [Help] [answered]
I've managed to build a from-scratch docker container in dagger, but it feels a bit hacky.

I defined scratch by using a `FROM SCRATCH` repo that I pushed to docker hub:
```cue
#Scratch:{
    build: {
        docker.#Build&{
            steps: [
                docker.#Pull&{
                    source: "jcvz/scratch"
                }
            ]
        }
    }
    output: build.output
}
```

Then, in the full build, I call it as the first build step:
```cue
_build_docker: docker.#Build&{
	steps: [
		#Scratch&{},
		docker.#Copy & {
			contents: _build_exe.export.directories."/server"
			dest: "/"
		},
		...
	]
}
```

I am wondering if there is some better / standardized way to do this. I tried some various blank docker images/builds, but got errors such as "rootfs not allowed" or "fs is not set".

**Comment by shykes** [comment]
`docker.#Pull` returns a `docker.#Image`, so you can call it directly:

```
#Scratch: docker.#Pull & {
  source: "jcvz/scratch"
}
```

**Reply by jchvz** [reply]
I didn't know that! Thanks for the quick reply + info.

**Reply by shykes** [reply]
I realize now that I just made your workaround simpler, which is probably not what you were looking for ;)

**Reply by jchvz** [reply]
Haha, it's ok, I still learned from it!

**Comment by helderco** [comment, marked as answer]
We could have a `docker.#Scratch`:

```cue
#Scratch: #Image & {
    rootfs: dagger.#Scratch
    config: {}
}
```

But in the meantime you can easily do this:

```cue
_build_docker: docker.#Build&{
	steps: [
		{ output: rootfs: dagger.#Scratch },
		docker.#Copy & {
			contents: _build_exe.export.directories."/server"
			dest: "/"
		},
		...
	]
}
```

Or this:

```cue
_build_docker: docker.#Build&{
	steps: [
		docker.#Copy & {
			// input: docker.#Scratch
			input: rootfs: dagger.#Scratch
			contents: _build_exe.export.directories."/server"
			dest: "/"
		},
		...
	]
}
```

**Reply by jchvz** [reply]
Thanks a lot. This is the kind of shorcut I was looking for.
